sql,cli: add payloads_for_trace builtin

angelapwen opened the pull request:
Possibly the final step to #55733. Resolves #58608 😄 

Previously it was quite cumbersome to view all payloads for a given
trace: we needed to join on the `node_inflight_trace_spans` vtable
to filter for span IDs that match a trace ID, then apply the
`payloads_for_span()` builtin to each span ID. This patch adds
syntactic sugar to the above query.

Instead of

```
WITH spans AS (
  SELECT span_id
  FROM crdb_internal.node_inflight_trace_spans
  WHERE trace_id = $TRACE_ID)
) SELECT *
  FROM spans, LATERAL crdb_internal.payloads_for_span(spans.span_id);
```

we can now simply use:
```
crdb_internal.payloads_for_trace($TRACE_ID);
```

and achieve the same result. The patch also adds all payloads for all
long-running spans to the `crdb_internal.node_inflight_trace_spans`
table of the debug.zip file.

Release note (sql change): Add `payloads_for_trace()` builtin so that
all payloads attached to all spans for a given trace ID will be
displayed, utilizing the `crdb_internal.payloads_for_span()`
builtin under the hood. All payloads for long-running spans are also
added to debug.zip in the `crdb_internal.node_inflight_trace_spans`
table dump.

Co-authored-by: Tobias Grieger <tobias.b.grieger@gmail.com>

Release justification: This patch is safe for release because it
adds syntactic sugar to an internal observability feature.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60922)
<!-- Reviewable:end -->

angelapwen commented:
___
*[pkg/sql/logictest/testdata/logic_test/contention_event, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MU984oI5M1grCd5Chqz:-MU984oJ7hu_4mkqcY2d:b2lqdzc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2422aa5b9de1f3b3c184bc09b0d3bfa06da2ac2a/pkg/sql/logictest/testdata/logic_test/contention_event#L73)):*
> ```
> SELECT * FROM crdb_internal.trace_id()
> 
> # SHOW PAYLOADS FOR TRACE is syntactic sugar for most of the above statement.
> ```

I confirmed that the `SHOW PAYLOADS FOR TRACE` syntax works on cockroach demo, but am having trouble testing it here. The logic test doesn't like the syntax with `$curr_trace_id` because I believe the `let $curr_trace_id` always parses `$curr_trace_id` to a string whereas the parser expects a signed int 64. 

Is there a better way to pass around an int in a logic test? Or (this seems hacky because I feel the only purpose would be for this logic test?) I could have the parser accept a string expression and attempt to convert into an int64, returning an error if it cannot. 


<!-- Sent from Reviewable.io -->

tbg commented:
I'm surprised to see top-level grammar added for what is essentially still an internal debugging feature. I was thinking we'd be able to add something more low-level that has the same ease of UX (`crdb_internal.payloads_for_trace` if that's possible for example)

irfansharif commented:
yea, ditto what Tobi said. I don't think we want to go as far as introducing a top-level grammar and constrain ourselves to backwards compatibility and making sure it's exposed as such in our docs. How about a `crdb_internal.payloads_for_trace` instead?

angelapwen commented:
> I'm surprised to see top-level grammar added for what is essentially still an internal debugging feature. I was thinking we'd be able to add something more low-level that has the same ease of UX (`crdb_internal.payloads_for_trace` if that's possible for example)

> How about a crdb_internal.payloads_for_trace instead?

This is definitely possible! (though I'm not actually sure how to make one built-in execute the result of another — there might be quite a lot of repeated code between `crdb_internal.payloads_for_trace` and `crdb_internal.payloads_for_span` that I'd probably pull out into a helper.

I went with top-level grammar because @knz had recommended using the SQL delegate package to abstract away a more complex query. I think all the [statements in that package](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/delegate/delegate.go) represent top-level grammar, but perhaps I misinterpreted his suggestion. I'll wait for Raphael's thoughts to clarify!

Commit:
sql,cli: add payloads_for_trace builtin

Previously it was quite cumbersome to view all payloads for a given
trace: we needed to join on the `node_inflight_trace_spans` vtable
to filter for span IDs that match a trace ID, then apply the
`payloads_for_span()` builtin to each span ID. This patch adds
syntactic sugar to the above query.

Instead of

```
WITH spans AS (
  SELECT span_id
  FROM crdb_internal.node_inflight_trace_spans
  WHERE trace_id = $TRACE_ID)
) SELECT *
  FROM spans, LATERAL crdb_internal.payloads_for_span(spans.span_id);
```

we can now simply use:
```
crdb_internal.payloads_for_trace($TRACE_ID);
```

and achieve the same result. The patch also adds all payloads for all
long-running spans to the `crdb_internal.node_inflight_trace_spans`
table of the debug.zip file.

Release note (sql change): Add `payloads_for_trace()` builtin so that
all payloads attached to all spans for a given trace ID will be
displayed, utilizing the `crdb_internal.payloads_for_span()`
builtin under the hood. All payloads for long-running spans are also
added to debug.zip in the `crdb_internal.node_inflight_trace_spans`
table dump.

Co-authored-by: Tobias Grieger <tobias.b.grieger@gmail.com>

Release justification: This patch is safe for release because it
adds syntactic sugar to an internal observability feature.

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60922)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, and @tbg)

___
*[pkg/sql/logictest/testdata/logic_test/contention_event, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MU984oI5M1grCd5Chqz:-MU9mHxJ8daIXhEZXlCy:bgyvcxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2422aa5b9de1f3b3c184bc09b0d3bfa06da2ac2a/pkg/sql/logictest/testdata/logic_test/contention_event#L73)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

I confirmed that the `SHOW PAYLOADS FOR TRACE` syntax works on cockroach demo, but am having trouble testing it here. The logic test doesn't like the syntax with `$curr_trace_id` because I believe the `let $curr_trace_id` always parses `$curr_trace_id` to a string whereas the parser expects a signed int 64. 

Is there a better way to pass around an int in a logic test? Or (this seems hacky because I feel the only purpose would be for this logic test?) I could have the parser accept a string expression and attempt to convert into an int64, returning an error if it cannot. 
</blockquote></details>

This works now :smile_cat: I had some errors with parsing the int previously. 


<!-- Sent from Reviewable.io -->

tbg commented:
Looks like we're mostly waiting for @knz here, assigning as sole reviewer to make that clearer (having everyone as reviewer can diffuse responsibility - I wonder if we should have one reviewer and the rest as assignees or simply CC'ed? :thinking:)

knz reviewed (commented):
To summarize what angela  and I talked about today:

1. agreed that we may not need a full-blown SHOW syntax form for this just yet.  We can build a crdb_internal function, also a SRF since it would return multiple rows, taking a trace ID as argument. We can build a SRF thanks to Yahor's recent investment in making the internal executor streamable. I advised to look at the code from `pg_builtins.go` as a source of inspiration.

2. hypothetically if we did decide to add a syntax form, I'd advise to ensure that it takes an arbitrary SQL scalar expression as argument, not just a numeric literal. This would make it possible to pass it the result of a function call or an arbitrary subquery. That is not relevant to this PR (see point above) but something to keep in mind.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-:-MUEEErx2LgxZCcPne7E:b-u6uwty)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen)



<!-- Sent from Reviewable.io -->

angelapwen commented:
This PR is ready for another review, implemented as a streaming built-in using `InternalExecutor` rather than top-level syntax. I have a question in the comments about the syntax of the inclusion in debug.zip which was requested in #58608. Thank you 🙏

tbg reviewed (commented):
Left some suggestions on the contention test and the debug zip test, I'll defer to Raphael on the actual impl of the builtin.

> Reviewed 13 of 13 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-:-MUJiSwH7NOkvSl8P0f2:bmj6adj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @irfansharif, @knz, and @tbg)

___
*[pkg/cli/zip.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJh9gm6yj9ZzJbGtW_:-MUJh9gm6yj9ZzJbGtWa:b-b81gda) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/cli/zip.go#L43)):*
> ```Go
> var customSelectClause = map[string]string{
> 	// TODO(angelapwen): How can I test this? All tests in zip_test pass but not sure this statement works appropriately.
> 	"crdb.internal.node_inflight_trace_spans": "*, LATERAL crdb_internal.payloads_for_span(span_id) WHERE duration > 10*time.Second ORDER BY trace_id ASC, duration DESC",
> ```

Hmm, yeah, this seems like a signal that the simple approach to specify a select clause isn't flexible enough. Maybe this should map table to replacement query?

So in this code

https://github.com/cockroachdb/cockroach/blob/14ab4b0b9aab9f74f6fbfa3e244dc337a05e663a/pkg/cli/zip_cluster_wide.go#L111-L115

Instead of looking up the selectClause, you'd have

```go
query := fmt.Sprintf(`SELECT * FROM %s`, table)
if override, ok := customQuery[table]; ok {
  query = override
}
```

You'd have to adapt dumpTableDataForZip accordingly here:

https://github.com/cockroachdb/cockroach/blob/14ab4b0b9aab9f74f6fbfa3e244dc337a05e663a/pkg/cli/zip.go#L235-L234

To do

```go
query = fmt.Sprintf(`SET statement_timeout = '%s'; `, zc.timeout) + query
```

(Where `query` is now the argument to the method formerly known as `selectClause`).

@knz does this check out?
___
*[pkg/sql/logictest/testdata/logic_test/contention_event, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJgaCQET8rj8ezgw3E:-MUJgaCQET8rj8ezgw3F:b-u198hb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/sql/logictest/testdata/logic_test/contention_event#L35)):*
> ```
> BEGIN;
> SET TRANSACTION PRIORITY HIGH;
> SELECT * FROM kv
> ```

Add a scan order here (ORDER BY x ASC) to ensure you always run into one intent first
___
*[pkg/sql/logictest/testdata/logic_test/contention_event, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJftdN9xxgRE3FR3to:-MUJftdN9xxgRE3FR3tp:b-eizgp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/sql/logictest/testdata/logic_test/contention_event#L62)):*
> ```
> ) SELECT COUNT(*) > 0
>   FROM payloads
>   WHERE payload_type = 'roachpb.ContentionEvent' AND crdb_internal.pretty_key(decode(payload_jsonb->>'key', 'base64'), 1) LIKE '/1/"k"/%'
> ```

Can you use LIMIT 1 instead of COUNT(*) so that the key gets printed?
I'm still interested in resolving the nondeterminism in the output. I think the root problem is that there are sometimes two contention events. The conflicting transaction lays down two intents, and you definitely block on the first. However, when the first is resovled, you may or may not still run into the second one for a short period of time (or so I am guessing based on the flake!). If no ordering is enforced, you can end up seeing either the first or the second here. The solution then should be: specify an ordering to the scan (see comment above) and use LIMIT 1 here. Then you should see the same key, always, and can print it in the result.
___
*[pkg/sql/logictest/testdata/logic_test/contention_event, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJgp1vBnktdKHH3v9P:-MUJgp1vBnktdKHH3v9Q:b-rceiqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/sql/logictest/testdata/logic_test/contention_event#L74)):*
> ```
>   FROM payloads
>   WHERE payload_type = 'roachpb.ContentionEvent'
>   AND crdb_internal.pretty_key(decode(payload_jsonb->>'key', 'base64'), 1) LIKE '/1/"k"/%'
> ```

Assuming the above works, please update this one as well.


<!-- Sent from Reviewable.io -->

angelapwen commented:
___
*[pkg/sql/logictest/testdata/logic_test/contention_event, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJftdN9xxgRE3FR3to:-MUJkgT4E68HMvq3D5wS:b-6p3sev) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/sql/logictest/testdata/logic_test/contention_event#L62)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Can you use LIMIT 1 instead of COUNT(*) so that the key gets printed?
I'm still interested in resolving the nondeterminism in the output. I think the root problem is that there are sometimes two contention events. The conflicting transaction lays down two intents, and you definitely block on the first. However, when the first is resovled, you may or may not still run into the second one for a short period of time (or so I am guessing based on the flake!). If no ordering is enforced, you can end up seeing either the first or the second here. The solution then should be: specify an ordering to the scan (see comment above) and use LIMIT 1 here. Then you should see the same key, always, and can print it in the result.
</blockquote></details>

I just gave this a quick try (with ordering the SELECT by k ASC above and using LIMIT 1) here. It still fails a stress test, alternately surfacing /1/"k"/1/1 and 1/"k"/0 as contention keys. I could technically always order these 2 keys in _this_ statement and limit to the first one, so it would always surface the alphabetically "earlier" key ie. /1/"k"/0 but I'm not sure that's what you had wanted. 


<!-- Sent from Reviewable.io -->

angelapwen commented:
___
*[pkg/sql/logictest/testdata/logic_test/contention_event, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJftdN9xxgRE3FR3to:-MUJthbM5mjGga9cK5T4:b-vzxo0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/sql/logictest/testdata/logic_test/contention_event#L62)):*
> If no ordering is enforced, you can end up seeing either the first or the second here. 

I also wanted to mention that I've seen both printed (when I print keys that match on '/1/k/&') in stress testing so it's not just either-or but there are situations where both contention events are present. 


<!-- Sent from Reviewable.io -->

angelapwen commented:
___
*[pkg/sql/logictest/testdata/logic_test/contention_event, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJftdN9xxgRE3FR3to:-MUJuL0MAUCbU68k-L-m:b-ue0rsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/sql/logictest/testdata/logic_test/contention_event#L62)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

> If no ordering is enforced, you can end up seeing either the first or the second here. 

I also wanted to mention that I've seen both printed (when I print keys that match on '/1/k/&') in stress testing so it's not just either-or but there are situations where both contention events are present. 
</blockquote></details>

% not & :smiley_cat: 


<!-- Sent from Reviewable.io -->

angelapwen commented:
___
*[pkg/cli/zip.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJh9gm6yj9ZzJbGtW_:-MUKbWD1At7lDgQf4ATR:b-7j2zh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/cli/zip.go#L43)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Hmm, yeah, this seems like a signal that the simple approach to specify a select clause isn't flexible enough. Maybe this should map table to replacement query?

So in this code

https://github.com/cockroachdb/cockroach/blob/14ab4b0b9aab9f74f6fbfa3e244dc337a05e663a/pkg/cli/zip_cluster_wide.go#L111-L115

Instead of looking up the selectClause, you'd have

```go
query := fmt.Sprintf(`SELECT * FROM %s`, table)
if override, ok := customQuery[table]; ok {
  query = override
}
```

You'd have to adapt dumpTableDataForZip accordingly here:

https://github.com/cockroachdb/cockroach/blob/14ab4b0b9aab9f74f6fbfa3e244dc337a05e663a/pkg/cli/zip.go#L235-L234

To do

```go
query = fmt.Sprintf(`SET statement_timeout = '%s'; `, zc.timeout) + query
```

(Where `query` is now the argument to the method formerly known as `selectClause`).

@knz does this check out?
</blockquote></details>

Makes sense, I've been playing around with that code too. I think it might be better to include it in`zip_per_node.go` versus `zip_cluster_wide.go` because this is a node-local vtable though? 

https://github.com/cockroachdb/cockroach/blob/3bae09b5d473e65479c68a221c05d3260cb31f4b/pkg/cli/zip_per_node.go#L198-L206


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60922)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @irfansharif, @knz, and @tbg)

___
*[pkg/cli/zip.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJh9gm6yj9ZzJbGtW_:-MUNkIFD-fH9MofkHBS2:bxofvzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/cli/zip.go#L43)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Makes sense, I've been playing around with that code too. I think it might be better to include it in`zip_per_node.go` versus `zip_cluster_wide.go` because this is a node-local vtable though? 

https://github.com/cockroachdb/cockroach/blob/3bae09b5d473e65479c68a221c05d3260cb31f4b/pkg/cli/zip_per_node.go#L198-L206
</blockquote></details>

I made these changes to both `zip_cluster_wide.go` and `zip_per_node.go` because they both use `dumpTableDataForZip`. This looks right to me. 

I'm still not entirely sure how to test that we are dumping output appropriately but `TestZipContainsAllInternalTables` passes and many of the other tests leverage the `systems.job` table that I also modified here (and pass). 


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> this looks nice. Some nits below.

> Reviewed 2 of 13 files at r2, 5 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-:-MUOe0cmBD3ez-syxh9n:b1kxwit)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @angelapwen, @irfansharif, and @tbg)

___
*[pkg/cli/zip.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJh9gm6yj9ZzJbGtW_:-MUOd9u25o9OZrN1W3Zf:bpdibm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/cli/zip.go#L43)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

I made these changes to both `zip_cluster_wide.go` and `zip_per_node.go` because they both use `dumpTableDataForZip`. This looks right to me. 

I'm still not entirely sure how to test that we are dumping output appropriately but `TestZipContainsAllInternalTables` passes and many of the other tests leverage the `systems.job` table that I also modified here (and pass). 
</blockquote></details>

Won't hurt to go and manually look into a generated debug.zip archive and verify manually there's some trace data in there.
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 1574 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUOdavu3sx4Rmsf7t5X:-MUOdavv5uHqNWBqPm7z:bklnhnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c521cce68ecde418f1b43d6c0cfe96759547ae/pkg/sql/sem/builtins/generator_builtins.go#L1574)):*
> ```Go
> 						(SELECT span_id
>   	 					FROM crdb_internal.node_inflight_trace_spans
>  		 					WHERE trace_id = %d
> ```

change that %d to `$1`

and remove the `Sprintf` call, make the query a `const` instead.
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 1585 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUOdcdv4tVlSaWYaP90:-MUOdcdv4tVlSaWYaP91:b-7ehs90) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c521cce68ecde418f1b43d6c0cfe96759547ae/pkg/sql/sem/builtins/generator_builtins.go#L1585)):*
> ```Go
> 		ctx.Txn,
> 		sessiondata.NoSessionDataOverride,
> 		query,
> ```

then pass the trace ID as argument at the end here.
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 1618 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUOdnALBRD5_bfQkUud:-MUOdnALBRD5_bfQkUue:b-4vtl8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c521cce68ecde418f1b43d6c0cfe96759547ae/pkg/sql/sem/builtins/generator_builtins.go#L1618)):*
> ```Go
> 	err := p.it.Close()
> 	if err != nil {
> 		// TODO(angelapwen): Ok to do this without alerting the iterator is still open?
> ```

This is a problem. 
Can you file an issue about this, and refer to the issue here.
Assign the issue to @yuzefovich 


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60922)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @angelapwen, @irfansharif, @knz, @tbg, and @yuzefovich)

___
*[pkg/cli/zip.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJh9gm6yj9ZzJbGtW_:-MUOzOED8_tiu0oa7Alx:bq0efp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/cli/zip.go#L43)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Won't hurt to go and manually look into a generated debug.zip archive and verify manually there's some trace data in there.
</blockquote></details>

Working on this!
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 1574 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUOdavu3sx4Rmsf7t5X:-MUOw37h7ypSHisBmaUT:b-k1v25) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c521cce68ecde418f1b43d6c0cfe96759547ae/pkg/sql/sem/builtins/generator_builtins.go#L1574)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

change that %d to `$1`

and remove the `Sprintf` call, make the query a `const` instead.
</blockquote></details>

Done. But wondering why this is preferable to the`Sprintf` call — is it that much cheaper performance- or memory-wise?
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 1585 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUOdcdv4tVlSaWYaP90:-MUOw-ZL91kr4mKH2Fhe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c521cce68ecde418f1b43d6c0cfe96759547ae/pkg/sql/sem/builtins/generator_builtins.go#L1585)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

then pass the trace ID as argument at the end here.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 1618 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUOdnALBRD5_bfQkUud:-MUOzAeR1vPFvm7tcpuJ:b13o1lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c521cce68ecde418f1b43d6c0cfe96759547ae/pkg/sql/sem/builtins/generator_builtins.go#L1618)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This is a problem. 
Can you file an issue about this, and refer to the issue here.
Assign the issue to @yuzefovich 
</blockquote></details>

Done. See #61123. I have assigned this todo to both myself and Yahor. 


<!-- Sent from Reviewable.io -->

angelapwen commented:
___
*[pkg/cli/zip.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60922#-MUJh9gm6yj9ZzJbGtW_:-MUP7FT7CThO9spOMrz6:b-uybcxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb6b41783e465292f60e25d036cff378f4a4362c/pkg/cli/zip.go#L43)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Working on this!
</blockquote></details>

Good thing I checked manually, there was a typo in my table name and it was defaulting to the `SELECT * FROM ...` syntax!

Anyway, now I have the table printing with these columns:
```
trace_id	parent_span_id	span_id	goroutine_id	finished	start_time	duration	operation	payload_type	payload_jsonb
```
Which I think is quite nice :sunflower: 


<!-- Sent from Reviewable.io -->

angelapwen commented:
Discussed with @tbg that I will file another issue for the non-deterministic output in the contention key and get this merged first 🏃‍♀️ 

bors=knz

angelapwen commented:
Discussed with @tbg that I will file another issue for the non-deterministic output in the contention key and get this merged first 🏃‍♀️ 

bors r=knz

angelapwen commented:
Note that when #61407 is resolved, this builtin will return redundant payloads. At the moment this builtin works as expected: it returns a row for each payload in all spans of a trace. This is because we are iterating through all tracked _root spans_  by way of the other `payloads_for_span` builtin and pulling all payloads of descendant spans from the recordings of the root spans. 

When we have a centralized span registry, then `payloads_for_span` will be able to pull payloads of any span, not just root spans. Therefore in this current implementation, a span's payload(s) would then be surfaced once for that span, and then one more time for each ancestor span.
